python/src/uagents/communication.py

```diff
@@ -0,0 +1,282 @@
+from uagents.utils import get_logger
+
+LOGGER = get_logger("dispenser", logging.DEBUG)
+
+
+class DeliveryStatus(str, Enum):
+    """Delivery status of a message."""
+
+    SENT = "sent"
+    DELIVERED = "delivered"
+    FAILED = "failed"
+
+
+@dataclass
+class MsgDigest:
+    """
+    Represents a message digest containing a message and its schema digest.
+
+    Attributes:
+        message (Any): The message content.
+        schema_digest (str): The schema digest of the message.
+    """
+
+    message: Any
+    schema_digest: str
+
+
+@dataclass
+class MsgStatus:
+    """
+    Represents the status of a sent message.
+
+    Attributes:
+        status (str): The delivery status of the message {'sent', 'delivered', 'failed'}.
+        detail (str): The details of the message delivery.
+        destination (str): The destination address of the message.
+        endpoint (str): The endpoint the message was sent to.
+    """
+
+    status: DeliveryStatus
+    detail: str
+    destination: str
+    endpoint: str
+
+
+class Dispenser:
+    """
+    Dispenses messages externally.
+    """
+
+    def __init__(self):
+        self._envelopes: List[Tuple[Envelope, List[str], asyncio.Future, bool]] = []
+        self._timeout: float = DEFAULT_DISPENSER_INTERVAL
+
+    def configure(self, timeout: Optional[float] = None):
+        """
+        Configure the dispenser.
+
+        Configuration may not be necessary for all implementations.
+        """
+        if timeout:
+            self._timeout = timeout
+
+    def add_envelope(
+        self,
+        envelope: Envelope,
+        endpoints: List[str],
+        response_future: asyncio.Future,
+        sync: bool = False,
+    ):
+        """
+        Add an envelope to the dispenser.
+
+        Args:
+            envelope (Envelope): The envelope to send.
+            endpoints (List[str]): The endpoints to send the envelope to.
+            response_future (asyncio.Future): The future to set the response on.
+            sync (bool, optional): True if the message is synchronous. Defaults to False.
+        """
+        self._envelopes.append((envelope, endpoints, response_future, sync))
+
+    async def run(self):
+        """Run the dispenser routine."""
+        while True:
+            for env, endpoints, response_future, sync in self._envelopes:
+                try:
+                    result = await send_exchange_envelope(
+                        envelope=env,
+                        endpoints=endpoints,
+                        sync=sync,
+                    )
+                    response_future.set_result(result)
+                except Exception as err:
+                    LOGGER.error(f"Failed to send envelope: {err}")
+                finally:  # sending an envelope is only tried once
+                    self._envelopes.remove((env, endpoints, response_future, sync))
+            await asyncio.sleep(DEFAULT_DISPENSER_INTERVAL)
```

Comment:
Not sure there is any need to sleep here. Why not just let this run continuously?

Reply:
I started out with a timeout as I didn't want to unnecessarily take up resources. We can throw the whole part out as well as the config as that is currently not needed.